torch/csrc/jit/script/compiler.cpp

```diff
@@ -836,32 +836,194 @@ struct to_ir {
       , graph(method.graph())
       , def(def)
       , resolver(std::move(resolver_))
+      , self(self)
       , environment_stack(nullptr) {
     JIT_ASSERT(resolver);
     pushFrame(graph->block());
 
-    auto schema = extractSchemaFromDef(def, bool(self));
-
-    std::vector<Argument> arguments, returns; // for schema
-    // inputs
-    auto it = def.decl().params().begin();
-    auto end = def.decl().params().end();
     // Type annotations exclude explicitly typing the "self" parameter, so in the
     // case that this is a method with self we expect one fewer parameter annotation
     // than the number of parameters this Def takes.
     if (self && def.decl().params().size() == 0) {
       throw ErrorReport(def.decl().params().range()) << "methods must have a self argument";
     }
+
+    auto schema = extractSchemaFromDef(def);
+    std::vector<Argument> arguments = emitFormalArguments(self, schema);
+    // body
+    auto stmts = def.statements();
+    auto stmts_begin = stmts.begin();
+    auto stmts_end = stmts.end();
+    c10::optional<Return> return_stmt;
+    if (stmts_begin != stmts_end && (*std::prev(stmts_end)).kind() == TK_RETURN) {
+      --stmts_end;
+      return_stmt = Return(*stmts_end);
+    }
+    emitStatements(stmts_begin, stmts_end);
+    std::vector<Argument> returns = emitReturn(return_stmt, schema);
+
+    method.setSchema({def.name().name(), std::move(arguments), std::move(returns)});
+    // remove any uses of tuples that we inserted that are not needed
+    LowerSimpleTuples(graph);
+    ConstantPooling(graph);
+  }
+
+private:
+  Method& method;
+  std::shared_ptr<Graph> graph;
+  Def def;
+  Resolver resolver;
+  SugaredValuePtr self;
+  std::unordered_map<int64_t, Value*> integral_constants;
+  std::unordered_map<double, Value*> fp_constants;
+
+  // Singly-linked list of environments. This top element contains a member
+  // `next` that points to the most immediate enclosing scope's value.
+  std::shared_ptr<Environment> environment_stack;
+
+  void pushFrame(Block * b) {
+    environment_stack = std::make_shared<Environment>(method, resolver, b, environment_stack);
+  }
+  std::shared_ptr<Environment> popFrame() {
+    auto old_frame = environment_stack;
+    environment_stack = environment_stack->next;
+    return old_frame;
+  }
+
+  std::vector<IValue> evaluateDefaults(const SourceRange& r, const std::vector<Expr>& default_types, const std::vector<Expr>& default_exprs) {
+    std::vector<IValue> default_values;
+    if (default_exprs.size() > 0) {
```

Comment:
would prefer the inverse, i.e. `if (defaults_exprs.empty()) return defaultValues` to avoid nesting